usvm-ts/src/main/kotlin/org/usvm/machine/expr/ExprUtil.kt

```
@@ -248,3 +248,16 @@ fun TsContext.checkReadingInRange(
         blockOnFalseState = { throwException("Index out of bounds: $index, length: $length") }
     )
 }
+
+fun TsContext.checkLengthBounds(
+    scope: TsStepScope,
+    length: UExpr<TsSizeSort>,
+    maxLength: Int,
+): Unit? {
+    // Check that length is non-negative and does not exceed `maxLength`.
+    val condition = mkAnd(
+        mkBvSignedGreaterOrEqualExpr(length, mkBv(0)),
+        mkBvSignedLessOrEqualExpr(length, mkBv(maxLength))
+    )
+    return scope.assert(condition)
```

It is different conditions. Reading by a negative element should raise an exception, since it is probably an error. Reading by 1001 is not an error in all cases

---

I remember, we decided that reading out of bounds is equally wrong in both cases. If we had decided to allow reading indices larger than the length, then it would have already been implemented, as it would simply return undefined.

~~I do agree, however, that it should use `scope.fork` and throw an exception, instead of just asserting.~~

---

Note: there are two different functions above this, which are intended for array reading: `checkNegativeIndexRead` and `checkReadingInRange`. They fork and throw on out-of-bounds access.
The function `ensureLengthBounds` that you have commented in, is for another case: when we read a symbolic length of an array, we explicitly force this symbolic value to be in range 0..length. Forking and throwing exception here does not make sense.